## 1. Ejemplos de uso correcto de ciclos

Como ya vimos (casi) nunca debemos usar `break` en los ciclos. 


Veamos otro ejemplo, ahora con el `for`: supongamos  que queremos averiguar si un número `n` es primo o no. Una forma sencilla de hacerlo es ir dividiendo por todos los números de `2` hasta `n-1` y si encontramos que uno de esos número divide a `n` el número es compuesto, en caso contrario es primo. Una implementación rápida podría ser

In [3]:
def es_primo(n: int) -> bool:
    # pre: n > 0
    # post: devuelve True si n  es primo, en caso contrario devuelve False
    primo = True
    for i in range(2, n):
        if n % i == 0:
            primo = False
            print(i, 'es divisor de', n)
            break
    return primo

print(es_primo(563456345634567)) # False
# print(es_primo(563456345634569)) # True (tarda mucho)

3 es divisor de 563456345634567
False


Esta implementación es correcta y traduce eficientemente el algoritmo que habíamos pensado.  Sin embargo, "romper" un ciclo `for` con un `break` no es una práctica recomendada. En este caso, donde el código es muy sencillo, no habría problemas, pero para códigos más complejos,  como ya dijimos, puede traer problemas.

Una propuesta sin el uso de `break` puede ser la siguiente:

In [ ]:
def es_primo(n: int) -> bool:
    # pre: n > 0
    # post: devuelve True si n  es primo, en caso contrario devuelve False
    primo = True
    i = 2
    while i < n and primo == True:
        if n % i == 0:
            primo = False
        i += 1
    return primo

print(es_primo(563456345634567)) # False

Obeservemos que el `for` debió ser reemplazado por el `while` puesto que estamos en el caso de un ciclo que no sabemos cuantas repeticiones tendrá.

Hay muchas formas  de hacer este programa, veamos otra forma. Usualmente podemos usar la negación de la condición que disparaba el `break` como una condición del `while`:

In [ ]:
def es_primo(n: int) -> bool:
    # pre: n > 0
    # post: devuelve True si n  es primo, en caso contrario devuelve False
    i = 2
    while i < n and n % i != 0:
        i += 1
    if i < n:
        return False
    else: 
        return True
        
print(es_primo(563456345634567)) # False

O la forma más concisa,  que pareciera ser la mejor con este algoritmo en mente:

In [ ]:
def es_primo(n: int) -> bool:
    # pre: n > 0
    # post: devuelve True si n  es primo, en caso contrario devuelve False
    i = 2
    while i < n and n % i != 0:
        i += 1
    return i == n 

## 4. Método de Newton para la raíz cuadrada

El *método de Newton* o de *Newton-Raphson* es un algoritmo para encontrar aproximaciones de los ceros o raíces de una función real. Una explicación completa se puede encontrar en Wikipedia en el artículo [Método de Newton](https://es.wikipedia.org/wiki/M%C3%A9todo_de_Newton). 

Se puede utilizar el  método de Newton para encontrar una aproximación numérica de la raíz cuadrada de un número real. Observar que si $a >0$
$$
b = \sqrt{a}\quad \Leftrightarrow \quad b^2 = a \quad\Leftrightarrow\quad b^2-a=0.
$$
Es decir, $\sqrt{a}$ es raíz de la función $f(x) = x^2 -a$. El algorimo que se obtiene de aplicar el  método de Newton a $f$ se puede describir de la siguiente forma: 
1. Sea $x_0$ un valor inicial arbitrario ($>0$). 
2. Sea 
\begin{equation}
x_{n+1} = \frac12(x_n + \frac{a}{x_n}) \tag{*}
\end{equation}
para $n \ge 0$. 

3. Aplique la fórmula (*) desde $n=0$ en adelante hasta obtener un $x_n$ que se aproxime "lo suficiente".

En el ítem 3. la frase "lo suficiente" no es del todo clara y  trataremos de interpretarla a lo largo de la sección.

Veamos como funciona en algún caso particular, por ejemplo, si $a$ es $4$ y $x_0$ es $3$. De ahora en más, por ser más cómodo; llamaremos $x$  a $x_n$ (el valor original) e $y$ a $x_{n+1}$ (el valor que se obtiene), luego la primera aproximación a $\sqrt{4} =2$ es 


In [4]:
a = 4
x = 3
y = (1/2)*(x + a/x)
print(y)

2.1666666666666665


El resultado es `2.1666666666666665` que está más cerca que `3` de la respuesta correcta ($\sqrt{4} = 2$). Si nosotros repetimos el proceso con la nueva estimación, se acerca aún más:


In [5]:
x = y
y = (1/2)*(x + a/x)
print(y)

2.0064102564102564


De la celda anterior obtenemos `2.0064102564102564`. Después de algunas actualizaciones más, la estimación es muy precisa:

In [6]:
x = y
y = (1/2)*(x + a/x)
print(y)
x = y
y = (1/2)*(x + a/x)
print(y)

2.0000102400262145
2.0000000000262146


Finalmente,  con una aproximación más obtenemos el resultado exacto: 

In [7]:
x = y
y = (1/2)*(x + a/x)
print(y)

2.0


En general, no sabemos de antemano cuántos pasos se necesitan para llegar a la respuesta correcta, pero sabemos cuándo llegamos porque la estimación
deja de cambiar:

In [8]:
x = y
y = (1/2)*(x + a/x)
print(y)

2.0


Cuando `y == x`, podemos detenernos. Aquí hay un bucle que comienza
con una estimación inicial, `x`, y la mejora hasta que deja de cambiar. Podemos escribir la función `raiz_cuadrada()` de la siguiente forma:


In [6]:
def raiz_cuadrada(a: float) -> float:
    # pre: a > 0
    # post: devuelve y una aproximación de la raíz cuadrada de a
    x = a / 2 # aproximamos la raíz de a por a/2
    y = (1/2)*(x + a/x)
    while y != x:
        x = y # el último calculado pasa a ser el que se usa para calcular el próximo
        y = (1/2)*(x + a/x)
    return y

print(raiz_cuadrada(4), raiz_cuadrada(4)**2)
print(raiz_cuadrada(9), raiz_cuadrada(9)**2)
print(raiz_cuadrada(30), raiz_cuadrada(30)**2)
print(raiz_cuadrada(101), raiz_cuadrada(101)**2)
print(raiz_cuadrada(102), raiz_cuadrada(102)**2)

2.0 4.0
3.0 9.0
5.477225575051661 30.0
10.04987562112089 101.0
10.099504938362077 101.99999999999999


Para muchos de los valores de `a` esto funciona bien, pero en general no es recomendable probar la igualdad de dos `float`. Los valores de coma flotante son solo aproximadamente correctos: la mayoría de los números racionales, como $ 1/3 $, y los números irracionales, como $\sqrt{2}$, no se pueden representar exactamente con un `float`.

En lugar de comprobar si `x` e `y` son exactamente iguales, es más seguro usar la función `abs` que devuelve el valor absoluto de un número real:  detenemos el bucle cuando `abs(x - y)` es menor que un error que consideramos aceptable. Redefinamos la función:

In [7]:
def raiz_cuadrada(a: float) -> float:
    # pre: a > 0
    # post: devuelve y una aproximación de la raíz cuadrada de a
    x = a / 2 # aproximamos la raíz de a por a/2
    y = (1/2)*(x + a/x)
    error = 0.0001
    while abs(x - y) > error:
        x = y
        y = (1/2)*(x + a/x)
    return y

print(raiz_cuadrada(4), raiz_cuadrada(4)**2)
print(raiz_cuadrada(9), raiz_cuadrada(9)**2)
print(raiz_cuadrada(30), raiz_cuadrada(30)**2)
print(raiz_cuadrada(101), raiz_cuadrada(101)**2)
print(raiz_cuadrada(102), raiz_cuadrada(102)**2)

2.0 4.0
3.0000000000393214 9.000000000235929
5.477225575302472 30.0000000027475
10.049875621244148 101.00000000247745
10.099504938503184 102.0000000028502


Esta última sería la forma más correcta de escribir la función. La primera forma depende mucho de la implementación del compilador y se pueden obtener resultados inciertos.

## 5. Programación defensiva 

El primer paso para obtener las resultados correctos de nuestros programas es asumir que los errores sucederán y protegerse contra ellos. Esto se llama programación defensiva, y la forma más común de hacerlo es agregar aserciones a nuestro código para que se compruebe a sí mismo a medida que se ejecuta. Una *aserción* es simplemente una declaración de que algo debe ser cierto en un cierto punto de un programa. La sintaxis de la aserción es la siguiente

```
assert condición, 'Cadena  que se imprime cuando condición == False'
```



Cuando Python ve una aserción, evalúa la condición de la aserción. Si es cierta, Python no hace nada, pero si es falsa, Python detiene el programa inmediatamente e imprime el mensaje de error explicitado por el programador. Por ejemplo, este fragmento de código suma todos los números de una lista de números positivos:


In [ ]:
numeros = [1.5, 2.3, 0.7, 0.001, 4.4] # lista de números (veremos esto en profundidad más adelante)
total = 0.0
for n in numeros:
    assert n > 0.0, 'Los datos solo deben contener valores positivos (' + str(n) + ' <= 0).'
    total += n
print('El total es:', total)

En  caso que la lista de números contenga un número no positivo,  obtendremos un error por el `assert`:

In [ ]:
numeros = [1.5, 2.3, 0.7, -0.001, 4.4]
total = 0.0
for n in numeros:
    assert n > 0.0, 'Los datos solo deben contener valores positivos (' + str(n) + ' <= 0).'
    total += n
print('El total es:', total)

### Precondiciones y postcondiciones (revisado)
Repasemos el concepto de precondición y postcondición y veamos que en algunos casos pueden verificarse con el comando `assert`. Como ya vimos, las *precondiciones* y *postcondiciones* son anotaciones dentro del cuerpo de una función que nos sirven para saber como deben ser los parámetros que recibe la función y  que es lo que hace la función:

- *Precondición:* es algo que debe ser cierto cuando comienza la la función, para que esta funcione correctamente.
- *Postcondición:* es algo que la función garantiza cierto cuando la misma termina. 

Las precondiciones y postcondiciones son parte de la documentación del programa y suelen escribirse como comentarios formales cuando se puede, o en caso contrario informales.

En el caso que se puedan formalizar en Python,  se pueden reemplazar o completar con `assert`.

Veamos un ejemplo con el máximo común divisor. La función escrita en la siguiente celda de código determina  el máximo común divisor de `a` y `b` cuando `a` y `b` son enteros no negativos y uno de ellos es mayor que 0. 

El algoritmo se basa en la dos propiedades siguientes: si $x$, $y$  son enteros,  con $y$ no nulo,  entonces:
\begin{align}
\operatorname{mcd}(x, y) &= \operatorname{mcd}(x, y - x) \\
\operatorname{mcd}(0, y) &= y.
\end{align}

In [ ]:
# 1º versión. 
def mcd(a, b: int) -> int:
    x, y = min(a, b), max(a, b)
    while x != 0: # "mientras x distinto de 0" 
        x, y = min(x, y - x), max(x, y - x)
    return y

mcd(15, 24)
# mcd(-3, 2)

Esta función se comporta bien mientras `a` y `b` cumplan que son no negativos y al menos uno sea no nulo. Verificar, por ejemplo, que


```
mcd(-3, 2)
```
es una instrucción que no termina.

Agreguemos ahora precondiciones y postcondiciones de manera informal:




In [ ]:
# 2º versión. 
def mcd(a, b: int) -> int:
    # pre: a y b no negativos, al menos uno de ellos no nulo.
    # post: devuelve el mcd de a y b
    x, y = min(a, b), max(a, b)
    while x != 0: # "mientras x distinto de 0" 
      x, y = min(x, y - x), max(x, y - x)
    return y

En  la tercera versión la precondición se escribe de manera formal, pero sigue siendo un comentario:

In [ ]:
# 3º versión
def mcd(a, b: int) -> int:
    # pre: (a >= 0 and b >=0) and (a != 0 or b != 0).
    # post: devuelve el mdc de a y b
    x, y = min(a, b), max(a, b)
    while x != 0: # "mientras x distinto de 0" 
        x, y = min(x, y - x), max(x, y - x)
    return y

La tercera versión no difiere en nada, al tiempo de ejecutarse, a la segunda versión, por lo tanto existe la posiblidad de entrar en un ciclo infinito para ciertos valores de `a`  y `b`.

En  la tercera versión hemos formalizado la precondición en lenguaje Python y por lo tanto  la podemos formalizar programáticamente con un `assert`:

In [ ]:
# 4º versión
def mcd(a, b: int) -> int:
    # pre: (a >= 0 and b >=0) and (a != 0 or b != 0).
    assert type(a) == int and type(b) == int and (a >= 0 and b >=0) and (a != 0 or b != 0), "Los parámetros deben ser enteros no negativos, al menos uno de ellos no nulo"
    # post: devuelve el mdc de a y b
    x, y = min(a, b), max(a, b)
    while x != 0: # "mientras x distinto de 0" 
      x, y = min(x, y - x), max(x, y - x)
    return y

print(mcd(15, 24))
# print(mcd(-3, 2))

Esta cuarta versión produce un error y detiene el programa cuando `a` y `b` no cumplen la precondición. En  este caso el `assert` nos garantiza que no se producirá un ciclo infinito y por lo tanto juega un rol importante em el programa. 

En el caso de la función `mcd()` formalizar matemáticamente la  postcondición es posible, pero no es sencillo implementarla en Python. Por lo tanto la podemos dejar escrita de manera informal o la escribimos con el fomalismo de la matemática, pero sin hacer un `assert` para verificarla. 

Por otro  lado, hacer un `assert` para la postcondición no parece razonable, puesto que si la función está bien implementada la postcondición siempre debe cumplirse y si ocurriera en algún caso que no se cumple es porque la función está mal definida. 


In [ ]:
# 5º versión
def mcd(a, b: int) -> int:
    # pre: (a >= 0 and b >=0) and (a != 0 or b != 0).
    assert type(a) == int and type(b) == int and (a >= 0 and b >=0) and (a != 0 or b != 0), "Los parámetros deben ser enteros no negativos, al menos uno de ellos no nulo"
    # post: devuelve d tal que
    #       1) a % d == 0 and b % d == 0, y
    #       2) si c entero y (a % c == 0 and b % c == 0) => d % c == 0
    x, y = min(a, b), max(a, b)
    while x != 0: # "mientras x distinto de 0" 
      x, y = min(x, y - x), max(x, y - x)
    return y